File: backend/app/domain/predictions.py
"""Prediction payloads, as a tagged union keyed on question kind.

The same discriminated union the frontend declares and `openapi.yaml`
documents. Plain dataclasses, no Pydantic — `domain/` imports no framework
(AGENTS.md hard rule 4), and the API layer converts at its own boundary.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class SectionPick:
    """One quarter-finalist pick, bound to the section it answers for."""

    section_index: int
    player_id: str


@dataclass(frozen=True, slots=True)
class QuarterFinalPicks:
    picks: tuple[SectionPick, ...]
    kind: str = "QF_PICKS"

    @property
    def player_ids(self) -> tuple[str, ...]:
        return tuple(pick.player_id for pick in self.picks)


@dataclass(frozen=True, slots=True)
class SemiFinalPicks:
    player_ids: tuple[str, ...]
    kind: str = "SF_PICKS"


@dataclass(frozen=True, slots=True)
class FinalistPicks:
    player_ids: tuple[str, ...]
    kind: str = "FINALIST_PICKS"


@dataclass(frozen=True, slots=True)
class ChampionPick:
    player_id: str
    kind: str = "CHAMPION"


@dataclass(frozen=True, slots=True)
class UnderperformerPick:
    player_id: str
    kind: str = "UNDERPERFORMER"


@dataclass(frozen=True, slots=True)
class BreakoutPick:
    player_id: str
    kind: str = "BREAKOUT"


@dataclass(frozen=True, slots=True)
class GenericPlayer:
    player_id: str
    kind: str = "GENERIC_PLAYER"


@dataclass(frozen=True, slots=True)
class GenericMatchResult:
    winner_id: str
    set_score: str
    kind: str = "GENERIC_MATCH_RESULT"


@dataclass(frozen=True, slots=True)
class GenericInteger:
    value: int
    kind: str = "GENERIC_INTEGER"


@dataclass(frozen=True, slots=True)
class GenericChoice:
    option_id: str
    kind: str = "GENERIC_CHOICE"


PredictionPayload = (
    QuarterFinalPicks
    | SemiFinalPicks
    | FinalistPicks
    | ChampionPick
    | UnderperformerPick
    | BreakoutPick
    | GenericPlayer
    | GenericMatchResult
    | GenericInteger
    | GenericChoice
)

#: A settled answer takes the same shape as the prediction it is compared with.
QuestionOutcomePayload = PredictionPayload

File: frontend/src/services/types/predictions.ts
import type { PlayerId, PredictionId, ParticipationId, QuestionId, IsoDateTime } from './ids';
import type { SetScore } from './enums';

/**
 * Prediction payloads, as a discriminated union keyed on question kind
 * (AGENTS.md frontend conventions). Do not widen this to a permissive record
 * type: the discriminant is what lets a form, a scoring reason and an API
 * schema all agree on the shape of one answer.
 *
 * The six TYPED kinds mirror spec 5.1. The four GENERIC_* kinds mirror the
 * generic answer types from spec 4.5.3, one per `AnswerType`.
 */

/** One quarter-finalist pick, bound to the section it came from. */
export interface SectionPick {
  /** 1..8 — the section this pick answers for. */
  readonly sectionIndex: number;
  readonly playerId: PlayerId;
}

/** Exactly one player per section, eight in total (spec 5.3). */
export interface QuarterFinalPicksPayload {
  readonly kind: 'QF_PICKS';
  readonly picks: readonly SectionPick[];
}

/** Four, each drawn from the quarter-final picks (spec 5.3). */
export interface SemiFinalPicksPayload {
  readonly kind: 'SF_PICKS';
  readonly playerIds: readonly PlayerId[];
}

/** Two, each drawn from the semi-final picks (spec 5.3). */
export interface FinalistPicksPayload {
  readonly kind: 'FINALIST_PICKS';
  readonly playerIds: readonly PlayerId[];
}

/** One of the finalist picks (spec 5.3). */
export interface ChampionPayload {
  readonly kind: 'CHAMPION';
  readonly playerId: PlayerId;
}

/** An entrant seeded 10 or better (spec 5.3). */
export interface UnderperformerPayload {
  readonly kind: 'UNDERPERFORMER';
  readonly playerId: PlayerId;
}

/** An unseeded entrant (spec 5.3). */
export interface BreakoutPayload {
  readonly kind: 'BREAKOUT';
  readonly playerId: PlayerId;
}

/** Generic question, answer type PLAYER. */
export interface GenericPlayerPayload {
  readonly kind: 'GENERIC_PLAYER';
  readonly playerId: PlayerId;
}

/**
 * Generic question, answer type MATCH_RESULT. The set score must be legal for
 * the draw's match format; the set-score point is only awarded when the winner
 * is also correct (spec 5.2).
 */
export interface GenericMatchResultPayload {
  readonly kind: 'GENERIC_MATCH_RESULT';
  readonly winnerId: PlayerId;
  readonly setScore: SetScore;
}

/** Generic question, answer type INTEGER. */
export interface GenericIntegerPayload {
  readonly kind: 'GENERIC_INTEGER';
  readonly value: number;
}

/** Generic question, answer type CHOICE. `optionId` indexes `Question.options`. */
export interface GenericChoicePayload {
  readonly kind: 'GENERIC_CHOICE';
  readonly optionId: string;
}

export type PredictionPayload =
  | QuarterFinalPicksPayload
  | SemiFinalPicksPayload
  | FinalistPicksPayload
  | ChampionPayload
  | UnderperformerPayload
  | BreakoutPayload
  | GenericPlayerPayload
  | GenericMatchResultPayload
  | GenericIntegerPayload
  | GenericChoicePayload;

export type PredictionPayloadKind = PredictionPayload['kind'];

/**
 * A settled answer takes the same shape as the prediction it is compared
 * against, which is what keeps scoring a pure comparison.
 */
export type QuestionOutcomePayload = PredictionPayload;

export interface Prediction {
  readonly id: PredictionId;
  readonly participationId: ParticipationId;
  readonly questionId: QuestionId;
  readonly payload: PredictionPayload;
  readonly submittedAt: IsoDateTime | null;
  readonly updatedAt: IsoDateTime;
}
